packages/core/src/features/st-symbol.ts

```diff
@@ -56,22 +139,66 @@ export function addSymbol({
     safeRedeclare?: boolean;
     localName?: string;
 }) {
-    const stSymbolData = plugableRecord.getUnsafe(context.meta.data, dataKey);
+    const { byNS, byNSFlat, byType } = plugableRecord.getUnsafe(context.meta.data, dataKey);
     const name = localName || symbol.name;
-    const existingSymbol = stSymbolData[name];
-    if (existingSymbol && node && !safeRedeclare) {
-        context.diagnostics.warn(node, diagnostics.REDECLARE_SYMBOL(name), {
-            word: name,
-        });
+    const typeTable = byType[symbol._kind];
+    const nsName = NAMESPACES[symbol._kind];
+    if (node && name === `root` && nsName === `main` && byNSFlat[nsName][name]) {
+        context.diagnostics.warn(node, diagnostics.REDECLARE_ROOT(), { word: `root` });
+        return;
     }
-    stSymbolData[name] = symbol;
+    byNS[nsName].push({ name, symbol, ast: node, safeRedeclare });
+    byNSFlat[nsName][name] = symbol;
+    typeTable[name] = symbol;
     // deprecated
     ignoreDeprecationWarn(() => {
         context.meta.mappedSymbols[name] = symbol;
     });
 }
 
+export function reportRedeclare(context: FeatureContext) {
+    const { byNS } = plugableRecord.getUnsafe(context.meta.data, dataKey);
+    for (const symbols of Object.values(byNS)) {
+        const flat: Record<string, SymbolDeclaration[]> = {};
+        const collisions: Set<string> = new Set();
+        for (const symbolDecl of symbols) {
+            const { name, safeRedeclare } = symbolDecl;
+            flat[name] = flat[name] || [];
+            if (!safeRedeclare && flat[name].length) {
+                collisions.add(name);
+            }
+            flat[name].push(symbolDecl);
+        }
+        for (const name of collisions) {
+            for (const { safeRedeclare, ast } of flat[name]) {
+                if (!safeRedeclare && ast) {
+                    context.diagnostics.warn(ast, diagnostics.REDECLARE_SYMBOL(name), {
+                        word: name,
+                    });
+                }
```

Comment:
what is there isn't ast don't we still want to do something?

Reply:
The only case that we have at the moment is the initial root that is registered for all stylesheets. I'm not sure what would we report on if a definition is not from a specific node in the AST. It would be odd to mark the entire AST as erroneous. Leaving it like it is for the moment (not different from the way it was before).